src/InMemoryMessageRepository.php

```diff
@@ -44,12 +44,30 @@ public function persist(Message ...$messages)
 
     public function retrieveAll(AggregateRootId $id): Generator
     {
+        $lastMessage = null;
+
         foreach ($this->messages as $message) {
             if ($id->toString() === $message->header(Header::AGGREGATE_ROOT_ID)->toString()) {
                 yield $message;
+                $lastMessage = $message;
+            }
+        }
+
+        return $lastMessage instanceof Message ? $lastMessage->aggregateVersion() : 0;
+    }
+
+    public function retrieveAllAfterVersion(AggregateRootId $id, int $aggregateRootVersion): Generator
```

Comment:
I think you could get away with:

```php
<?php

$messages = $this->retrieveAll($id);

foreach ($messages as $message) {
	if ($message->header(Header::AGGREGATE_ROOT_VERSION) > $aggregateRootVersion) {
		yield $message
	}
}

return $messages->getReturn();

```

Reply:
Actually, probably not, there could be no `$lastMessage` here but one in `retrieveAll()`